File: node/src/output.ts
import type { ImageResult } from "./types.js";

export function formatJson(results: ImageResult[]): string {
  return JSON.stringify(results, null, 2);
}

export function formatText(results: ImageResult[]): string {
  return results
    .map((r) => {
      const commit = r.commit ? r.commit.slice(0, 12) : "-";
      const repo = r.repo?.replace("https://", "") ?? "-";
      const confidence = r.confidence ?? "-";
      return (
        `${r.service}: ${r.image}\n` +
        `  repo:       ${repo}\n` +
        `  commit:     ${commit}\n` +
        `  status:     ${r.status}\n` +
        `  confidence: ${confidence}` +
        (r.commit_url ? `\n  url:        ${r.commit_url}` : "") +
        (r.matched_tag ? `\n  note:       commit is from matched tag '${r.matched_tag}', not the exact image digest` : "")
      );
    })
    .join("\n\n");
}

File: python/code_provenance/output.py
import json
from dataclasses import asdict
from code_provenance.models import ImageResult


def format_json(results: list[ImageResult]) -> str:
    """Format results as a JSON array."""
    return json.dumps([asdict(r) for r in results], indent=2)


def format_text(results: list[ImageResult]) -> str:
    """Format results as plain text, one block per image."""
    blocks = []
    for r in results:
        commit = r.commit[:12] if r.commit else "-"
        repo = r.repo.replace("https://", "") if r.repo else "-"
        confidence = r.confidence or "-"
        lines = [
            f"{r.service}: {r.image}",
            f"  repo:       {repo}",
            f"  commit:     {commit}",
            f"  status:     {r.status}",
            f"  confidence: {confidence}",
        ]
        if r.commit_url:
            lines.append(f"  url:        {r.commit_url}")
        if r.matched_tag and r.resolution_method == "ghcr_cross_lookup":
            lines.append(f"  note:       commit is from matched tag '{r.matched_tag}', not the exact image digest")
        elif r.matched_tag:
            lines.append(f"  tag:        {r.matched_tag}")
        blocks.append("\n".join(lines))
    return "\n\n".join(blocks)
